### DISCUSSION TITLE
Is it possible to render a popover on text highlight?

### DISCUSSION THREAD

**Discussion by tanshunyuan** [floating-ui/floating-ui#2095] [Q&A] [answered]
Currently, I'm trying to render a popover when certain text are highlighted as shown in the video. 
https://user-images.githubusercontent.com/29978940/211035880-6ea61f30-7966-4398-a1fb-0ab86bf03a15.mov

However, I have no idea how to use the selected text as a reference point for the popover to display correctly. Here's my codesandbox for trying it out https://codesandbox.io/s/disecting-floating-popover-eyqrhw?file=/src/App.ts

Any ideas on how I can do it?

**Comment by atomiks** [comment, marked as answer]
You would use a [virtual element](https://floating-ui.com/docs/virtual-elements) and the [`positionReference`](https://floating-ui.com/docs/react#changing-the-positioning-reference-while-retaining-events) callback ref to pass in the virtual element.

**Reply by tanshunyuan** [reply]
Nice it works! I'm going to paste the working code here for future reference!

```js
import "./styles.css";
import { useEffect, useState } from "react";
import { useFloating } from "@floating-ui/react";

export interface HighlightPosition {
  pageWidth: number;
  pageHeight: number;
  rect: DOMRect;
}

export default function App() {
  const [open, setOpen] = useState(false);
  const [highlightPosition, setHighlightPosition] = useState<HighlightPosition>(
    {
      pageWidth: 0,
      pageHeight: 0,
      rect: new DOMRect()
    }
  );
  const getSelectedText = () => {
    const winSelection = window.getSelection();
    if (winSelection) {
      const rect = winSelection.getRangeAt(0).getBoundingClientRect();
      const { scrollWidth, scrollHeight } = window.document.body;

      setHighlightPosition({
        rect,
        pageWidth: scrollWidth,
        pageHeight: scrollHeight
      });

      setOpen((v) => !v);
    }
  };

  const { x, y, strategy, positionReference, floating } = useFloating({
    placement: "top",
    open,
    onOpenChange: setOpen
  });

  useEffect(() => {
    const virtualEl = {
      getBoundingClientRect() {
        console.log("virtualEl");
        return highlightPosition.rect;
      }
    };
    positionReference(virtualEl);
  }, [highlightPosition.rect, positionReference]);

  return (
    <div className="App">
      <h1 onMouseUp={getSelectedText}>
        Floating UI — separating the event reference from the positioning
        reference
      </h1>

      <div
        ref={floating}
        style={{
          position: strategy,
          left: x ?? 0,
          top: y ?? 0,
          background: "black",
          color: "white",
          padding: 5
        }}
      >
        Tooltip
      </div>
    </div>
  );
}

```

**Comment by joaom00** [comment]
Hey @atomiks, sorry for the ping :sweat_smile: 

Is there a way to keep the floating element attached to the virtual element (in this case it would be the text selection rect) while scrolling?

**Reply by atomiks** [reply]
Adding `whileElementsMounted: autoUpdate` in `useFloating`, and adding `contextElement: elementBeingTextSelected` to the virtual element should work

If it doesn't there are two alternatives:

1. Make sure your floating element is inside the same `offsetParent` as the text selection element. This will make it follow with the scroll without needing to be updated.
2. Make sure your virtual element text selection rect is being updated on every scroll event so the rect can be calculated each time.

**Reply by joaom00** [reply]
Thanks, it works only with the last one. Maybe I did something wrong with the other options :sweat_smile:

**Reply by atomiks** [reply]
I made an example here: https://codesandbox.io/s/floating-ui-react-range-selection-tkpj9v?file=/src/App.tsx

The key is to make sure that `getBoundingClientRect` and/or `getClientRects` is returning a fresh value, not a stale one (i.e. a variable assigned from outside).

Fresh:
```js
const range = selection?.getRangeAt(0);

if (range) {
  refs.setReference({
    getBoundingClientRect: () => range.getBoundingClientRect(),
    getClientRects: () => range.getClientRects(),
  });
}
```

Stale:
```js
const range = selection?.getRangeAt(0);

if (range) {
  const boundingRect = range.getBoundingClientRect();
  const rects = range.getClientRects();
  refs.setReference({
    getBoundingClientRect: () => boundingRect,
    getClientRects: () => rects,
  });
}
```

**Reply by joaom00** [reply]
Wow, thank you so much for the example!
